Spec: topdown spec (#920)
consensus-shipyard/ipc: specs/topdown.md

> #Topdown Finality

Reviewer: This line isn't rendered as a title, add a " "
```suggestion
# Topdown Finality
```

> The topdown finality propagates the states in the parent subnet to the child blockchain. The following data are passed:

Reviewer: ```suggestion
The topdown finality propagates the states in the parent subnet to the child blockchain. The following data is passed:
```

> The topdown finality propagates the states in the parent subnet to the child blockchain. The following data are passed:

Reviewer: I'd suggest to first describe what is the purpose of topdown finality in IPC, then when exactly it happens, and only then the exact data which is being passed from parent to the child. Wdyt?

- Parent block height + block hash
- Transactions from the parent to the child subnet
- Validator changes from parent to the child subnet

The finality is represented in contract in the form of:

```solidity
// Link: https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/contracts/src/structs/CrossNet.sol#L11
struct ParentFinality {
    uint256 height;
    bytes32 blockHash;
}
```

The validator changes are represented as an array of Changes while topdown messages are an array of [IPCEnvelope](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/contracts/src/structs/CrossNet.sol#L63).

From a high level point of view, the topdown finality works as follows:

> - Parent finality process will first fetch the last committed `ParentFinality`. If there is no previous committed parent finality, the genesis block is [used](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/mod.rs#L36).

Reviewer: Can you describe what and when triggers "Parent finality process"?

Author: There is actually no one triggering the process, it's always running in the background once enabled. I have added this into the spec.

> - There is a `[ParentSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/syncer.rs#L24C19-L24C36)` that constantly polls the parent states through RPC calls, either fetching events emitted or through getters. The `ParentSyncer` stores the pulled data in cache and publishes a vote on the latest block seen.

Reviewer: Let's remove ` to fix rendering of the link

```suggestion
- There is a [ParentSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/syncer.rs#L24C19-L24C36) that constantly polls the parent states through RPC calls, either fetching events emitted or through getters. The `ParentSyncer` stores the pulled data in cache and publishes a vote on the latest block seen.
```

- Once a quorum is formed on the blocks seen, a topdown finality proposal will be [added](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/interpreter/src/chain.rs#L132) to the cometbft proposal. The proposal is just the `ParentFinality` struct shown above, i.e. a block height and the corresponding block hash.
- Once a topdown proposal is received by each node, it will be validated against the topdown syncer cache and the RPC node if there is a cache miss. If the checks do not pass, the proposal will be rejected. The checks include:
    - The height proposed is present in the cache or exists in the RPC node

>     - The hash is matches the corresponding height’s block hash

Reviewer: ```suggestion
    - The hash matches the corresponding height’s block hash
```

- Once the proposal is accepted, it will be executed. This means the topdown messages will be executed and validator changes will be stored.

> There are some complexity in parent finality process due to each node might have a different RPC endpoint and each RPC might not see exactly the same historical data. Currently topdown parent finality trusts the parent RPC node completely. In this case, one needs a way to detect if the RPC endpoints are serving correct data and all nodes agree on the same parent state. This is why a separate voting process is introduced for parent finality.

Reviewer: ```suggestion
All nodes must agree on the parent state, even though they may be using different RPC endpoints. Those RPC endpoints may have different views of the parent state. At the moment the nodes don't have any way to verify the correctness of `ParentFinality` received from the RPC endpoints. That's why the nodes have to reach consensus on the parent finality through a voting mechanism.
```

# Parent Syncer

The parent syncer polls the parent to obtain the “finalized” state in the parent. The “finalized” view in the parent is by delaying a fixed number of blocks from the chain head. This ensures the probability of a parent reorg is small.

> There are currently two implementations of parent syncer, [`LotusParentSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/syncer.rs#L24)` and [`TendermintAwareSyncer`](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/tendermint.rs#L16C15-L16C36). The `TendermintAwareSyncer` is a wrapper of `LotusParentSyncer` , which stops the polling if the node is currently syncing with its peers, because if it is, then it won’t be proposing anything until it’s finished, and it will receive the parent finalities from the CometBFT blocks instead.

Reviewer: Fixed links 

```suggestion
There are currently two implementations of parent syncer, [LotusParentSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/syncer.rs#L24) and [TendermintAwareSynce`](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/tendermint.rs#L16C15-L16C36). The `TendermintAwareSyncer` is a wrapper of `LotusParentSyncer` , which stops the polling if the node is currently syncing with its peers, because if it is, then it won’t be proposing anything until it’s finished, and it will receive the parent finalities from the CometBFT blocks instead.
```

# Vote Tally

The [`VoteTally`](https://github.com/consensus-shipyard/ipc/blob/specs/fendermint/vm/topdown/src/voting.rs) is a component used to keep track of votes cast on parent finalities by the current child validators, and supports looking for the highest parent block height and hash which received a quorum. The votes are published over a GossipSub channel, once a node observes a finalized block through the syncer described above.

A quorum detected by the `VoteTally` is used a pre-condition for finality proposals being added to the CometBFT block proposals, to avoid any liveness issues which could arise if the other validators were to reject the proposal. If a premature finality causes the block propsal to fail, it means in that round CometBFT cannot make progress, it cannot finalize a block, potentially causing the subnet blockchain to stall. By requiring a quorum, we avoid this issue by only proposing when we have high confidence that the proposal will be accepted.

> See [IPC Spec - IPLD Resolver](https://www.notion.so/IPC-Spec-IPLD-Resolver-7b4290a0d60c40cdba98cd6d3e66648b?pvs=21) for a more detailed discussion of the `VoteTally`.

Reviewer: It shouldn't link to Notion anymore.

Author: I removed the link for now, will add them back once they are merged

# Topdown Finality Proposal Execution

Once a proposal is accepted, it will be [executed](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/interpreter/src/chain.rs#L290) together with the batch of transactions. The execution of the proposal happens in three stages:

- Commit new topdown finality
- Fetch validator changes and store
- Fetch topdown messages and execute

The validator changes and topdown messages are first fetched from the parent syncer cache. If they are not present, then it will fallback to the RPC node. There is a simple exponential backoff [retry](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/finality/fetch.rs#L174). If the retry fails, which means the node cannot be reached to obtain the validator changes or topdown messages, an error is thrown. The reason for this is that the cluster has decided on the topdown finality, if one cannot obtain the corresponding changes and message, then it should stop.

As topdown finality relies heavily on RPC node for querying, make sure the RPC node is reliable and returns enough historical data.

> See also [IPS Spec - Executions](https://www.notion.so/IPS-Spec-Executions-ebf13d833d6845ec9c11b59bd514fcda?pvs=21) for a description of how proposals and and executions are implemented.

Reviewer: ```suggestion
See also [IPS Spec - Executions](https://www.notion.so/IPS-Spec-Executions-ebf13d833d6845ec9c11b59bd514fcda?pvs=21) for a description of how proposals and executions are implemented.
```

> See also [IPS Spec - Executions](https://www.notion.so/IPS-Spec-Executions-ebf13d833d6845ec9c11b59bd514fcda?pvs=21) for a description of how proposals and and executions are implemented.

Reviewer: It shouldn't link to Notion anymore.

Author: the issue is currently there are no links to those specs yet, I will remove the hyper link first and add them back later.

- Once a topdown proposal is received by each node, it will be validated against the topdown syncer cache and the RPC node if there is a cache miss. If the checks do not pass, the proposal will be rejected. The checks include:
    - The height proposed is present in the cache or exists in the RPC node
    - The hash matches the corresponding height’s block hash
- Once the proposal is accepted, it will be executed. This means the topdown messages will be executed and validator changes will be stored.

All nodes must agree on the parent state, even though they may be using different RPC endpoints. Those RPC endpoints may have different views of the parent state. At the moment the nodes don't have any way to verify the correctness of `ParentFinality` received from the RPC endpoints. That's why the nodes have to reach consensus on the parent finality through a voting mechanism.

# Parent Syncer

The parent syncer polls the parent to obtain the “finalized” state in the parent. The “finalized” view in the parent is by delaying a fixed number of blocks from the chain head. This ensures the probability of a parent reorg is small.

> There are currently two implementations of parent syncer, [LotusParentSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/syncer.rs#L24) and [TendermintAwareSynce`](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/tendermint.rs#L16C15-L16C36). The `TendermintAwareSyncer` is a wrapper of `LotusParentSyncer` , which stops the polling if the node is currently syncing with its peers, because if it is, then it won’t be proposing anything until it’s finished, and it will receive the parent finalities from the CometBFT blocks instead.

Reviewer: ```suggestion
There are currently two implementations of parent syncer, [LotusParentSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/syncer.rs#L24) and [TendermintAwareSyncer](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/topdown/src/sync/tendermint.rs#L16C15-L16C36). The `TendermintAwareSyncer` is a wrapper of `LotusParentSyncer` , which stops the polling if the node is currently syncing with its peers, because if it is, then it won’t be proposing anything until it’s finished, and it will receive the parent finalities from the CometBFT blocks instead.
```
